File: icons.py
"""Inline SVG icon library for converted pages and the pack CSS."""
from __future__ import annotations

import json
from pathlib import Path

ROOT = Path(__file__).resolve().parent
_CATALOG: dict | None = None

TITLE_ICONS = (
    ("spend amount", "clock"),
    ("amount limit", "clock"),
    ("spend count", "clipboard"),
    ("count limit", "clipboard"),
    ("time restriction", "clock"),
    ("time-of-day", "clock"),
    ("date restriction", "calendar"),
    ("calendar date", "calendar"),
    ("acceptance method", "credit-card"),
    ("country code", "globe"),
    ("country", "globe"),
    ("merchant", "store"),
    ("spend", "sliders"),
    ("account holder", "user"),
    ("accountholder", "user"),
    ("organisation", "building"),
    ("organization", "building"),
    ("onboard", "user-plus"),
    ("card", "credit-card"),
    ("transaction", "receipt"),
    ("search", "search"),
    ("provision", "smartphone"),
    ("wallet", "wallet"),
    ("google", "smartphone"),
    ("apple", "smartphone"),
    ("account", "building"),
    ("api", "code"),
    ("architecture", "layers"),
    ("overview", "layers"),
    ("flow", "git-branch"),
)


def icon_catalog() -> dict:
    global _CATALOG
    if _CATALOG is None:
        _CATALOG = json.loads((ROOT / "data" / "icons.json").read_text(encoding="utf-8"))
    return _CATALOG


def icon_html(name: str, size: int = 22) -> str:
    spec = icon_catalog().get(name)
    if not spec:
        return ""
    return (
        f'<span class="pack-icon" data-icon="{name}" contenteditable="false" aria-hidden="true">'
        f'<svg width="{size}" height="{size}" viewBox="0 0 24 24" fill="none" '
        f'stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">'
        f'{spec["svg"]}</svg></span>'
    )


def icon_for_title(title: str) -> str:
    hay = (title or "").lower()
    for token, name in TITLE_ICONS:
        if token in hay:
            return name
    return "file-text"


def wrap_callout(inner: str, warning: bool = False) -> str:
    name = "warning" if warning else "info"
    cls = "callout-warning" if warning else "callout"
    return f'<div class="{cls} callout-with-icon">{icon_html(name)}<div>{inner}</div></div>'

File: js/icons.js
const IconLibrary = {
  items: {},

  async load() {
    if (Object.keys(this.items).length) return this.items;
    const res = await fetch("/data/icons.json");
    this.items = await res.json();
    return this.items;
  },

  html(name, size) {
    const spec = this.items[name];
    if (!spec) return "";
    const px = size || 22;
    return (
      `<span class="pack-icon" data-icon="${name}" contenteditable="false" aria-hidden="true">` +
      `<svg width="${px}" height="${px}" viewBox="0 0 24 24" fill="none" stroke="currentColor" ` +
      `stroke-width="2" stroke-linecap="round" stroke-linejoin="round">${spec.svg}</svg></span>`
    );
  },

  list() {
    return Object.entries(this.items).map(([name, spec]) => ({ name, ...spec }));
  },

  grouped(query) {
    const needle = (query || "").trim().toLowerCase();
    const groups = new Map();
    for (const item of this.list()) {
      const hay = `${item.name} ${item.label} ${item.group}`.toLowerCase();
      if (needle && !hay.includes(needle)) continue;
      if (!groups.has(item.group)) groups.set(item.group, []);
      groups.get(item.group).push(item);
    }
    return groups;
  },
};
